lib/Smr/Container/DiContainer.php

```diff
@@ -0,0 +1,112 @@
+			/*
+			 * mysqli is a 3rd party library, and we do not have control over its constructor.
+			 * In order for PHP-DI to construct an instance of a class, each constructor argument must
+			 * be able to be constructed.
+			 * PHP-DI cannot construct mysqli by itself, because all of its arguments are primitive types.
+			 * Therefore, we need to declare a provider factory for the container to use when constructing new instances.
+			 *
+			 * The factories themselves are able to use dependency injection as well, so we can provide the MysqlProperties
+			 * typehint to make sure the container constructs and instance and provides it to the factory.
+			 */
+			mysqli::class => function (MySqlProperties $mysqlProperties): mysqli {
+				// Set the mysqli driver to raise exceptions on errors
+				if (!mysqli_report(MYSQLI_REPORT_ERROR | MYSQLI_REPORT_STRICT)) {
+					throw new RuntimeException('Failed to enable mysqli error reporting');
+				}
+				return new mysqli(
+					$mysqlProperties->getHost(),
+					$mysqlProperties->getUser(),
+					$mysqlProperties->getPassword(),
+					$mysqlProperties->getDatabaseName());
+			},
+			Dotenv::class => function (): Dotenv {
+				return Dotenv::createArrayBacked(ROOT);
+			},
+			// Explicitly name all classes that are autowired, so we can take advantage of
+			// the compiled container feature for a performance boost
+			MySqlProperties::class => autowire(),
+			MySqlDatabase::class => autowire()
+		];
+	}
+
+	private function buildContainer(): Container {
+		$builder = new ContainerBuilder();
+		return $builder
+			->addDefinitions($this->getDefinitions())
+			->useAnnotations(false)
+			->useAutowiring(true)
+			// The CompiledContainer.php will be saved to the /tmp directory on the Docker container once
+			// during its lifecycle (first request)
+			->enableCompilation("/tmp")
+			->build();
+	}
+
+	/**
+	 * Create a new DI\Container instance.
+	 * This needs to be done once during a bootstrapping script, like htdocs/config.inc
+	 */
+	public static function initializeContainer() {
+		new DiContainer();
+	}
+
+	/**
+	 * Retrieve the managed instance of $className, or construct a new instance with all dependencies.
+	 * @param string $className The name of the class to retrieve from the container.
+	 * @return mixed
+	 * @throws \DI\DependencyException
+	 * @throws \DI\NotFoundException
+	 */
+	public static function get(string $className) {
+		return self::getContainer()->get($className);
+	}
+
+	/**
+	 * Construct a fresh instance of $className. Dependencies will be retrieved from the container if they
+	 * are already managed, and created themselves if they are not.
+	 * @param string $className The name of the class to construct.
+	 * @return mixed
+	 * @throws \DI\DependencyException
+	 * @throws \DI\NotFoundException
+	 */
+	public static function make(string $className) {
+		return self::getContainer()->make($className);
+	}
+
+	/**
+	 * Return the raw Container instance
+	 * @return Container
```

Comment:
I'd say let's get rid of all the docstring elements that offer no documentation above and beyond the interface itself.

Reply:
Hm, I mean I think it's good to document what it's actually returning. Like, what `Container`? An explanation of the container it's returning seems useful to me. Maybe reword it? Any thoughts?

Reply:
After doing some testing with phpdoc, I'm kind of annoyed that it doesn't infer the type from the typehint. I dislike duplicating programmatic information in the docstring, because it increases the cost of maintaining it and the likelihood of desynchronization. Anyway, that's not something we can solve here, so I'm happy with this docstring as-is. :)

Reply:
Are you talking about how auto completion isn't working when using this method or something?

Reply:
No, just the phpDocumentor web pages (https://smrealms.github.io/api-docs/).